addons/io/org.openhab.io.neeo/src/main/java/org/openhab/io/neeo/internal/NeeoDeviceKeys.java

```diff
@@ -0,0 +1,201 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.io.neeo.internal;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.apache.commons.httpclient.HttpStatus;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.io.neeo.NeeoConstants;
+import org.openhab.io.neeo.NeeoUtil;
+import org.openhab.io.neeo.ServiceContext;
+import org.openhab.io.neeo.internal.net.HttpRequest;
+import org.openhab.io.neeo.internal.net.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+/**
+ * Stores the mapping of {@link ThingUID} to all brain keys
+ *
+ * @author Tim Roberts - Initial contribution
+ */
+public class NeeoDeviceKeys {
+
+    /** The logger */
+    private final Logger logger = LoggerFactory.getLogger(NeeoApi.class);
+
+    /** The lock controlling access to uidToKey */
+    private final ReadWriteLock cacheLock = new ReentrantReadWriteLock();
+
+    /** The mapping between ThingUID to brain keys */
+    private final Map<ThingUID, Set<String>> uidToKey = new HashMap<ThingUID, Set<String>>();
+
+    /** The service context */
+    private final ServiceContext context;
+
+    /** The brain's url */
+    private final String brainUrl;
+
+    /**
+     * Creates the object from the context and brainUrl
+     *
+     * @param context the non-null context
+     * @param brainUrl the non-empty brain url
+     */
+    public NeeoDeviceKeys(ServiceContext context, String brainUrl) {
+        Objects.requireNonNull(context, "context cannot be null");
+        NeeoUtil.requireNotEmpty(brainUrl, "brainUrl cannot be empty");
+
+        this.context = context;
+        this.brainUrl = brainUrl;
+    }
+
+    /**
+     * Refreshes the keys from the brain
+     *
+     * @throws IOException Signals that an I/O exception has occurred.
+     */
+    public void refresh() throws IOException {
+        final Lock writerLock = cacheLock.writeLock();
+        writerLock.lock();
+        try (HttpRequest request = new HttpRequest()) {
+            logger.debug("Getting existing device mappings from {}{}", brainUrl, NeeoConstants.PROJECTS_HOME);
+            final HttpResponse resp = request.sendGetCommand(brainUrl + NeeoConstants.PROJECTS_HOME);
+            if (resp.getHttpCode() != HttpStatus.SC_OK) {
+                throw resp.createException();
+            }
+
+            uidToKey.clear();
+
+            final JsonParser parser = new JsonParser();
+            final JsonObject root = parser.parse(resp.getContent()).getAsJsonObject();
```

Comment:
This looks like traversing manually through the json object. Isn't it possible to create a data object, deserialize it and then access the fields directly?

Reply:
Same comment as above - since I'm using internalish APIs, the 'structure' of the result has changed a bunch and I didn't really want to codify it all since all I care about is the key and adapter name.  I guess I could create a bare minimum structure to do the work...

Reply:
Just took a look at this and the second reason is that they property key names are dynamic (poor choice on their part).  Here's a small example - they use the name of the room as the key names:

	"name": "home",
	"rooms": {
		"Living Room": {
			"name": "Living Room",
			"hasController": false,
			"key": "6296031288356765696"
		},
		"Bedroom": {